Register Attributes Without Changes to ZEND APIs

kooldev opened the pull request:
This PR avoids modifications to existing Zend APIs by providing attribute registration functions. This was suggested by @nikic in the attributes v2 PR. All the work is done by `zend_add_attribute()` and some very small inline helper functions have been added to `zend_attributes.h` to deal with most common use cases.

The compiler has been adjusted make use of the new API. The changes ensure that all attributes (and arrays) are allocated in `zend_attributes.c`. This keeps the code in one place. It also enabled me to erase `zend_attribute_free()` from the public API because it is only needed internally.

kooldev committed:
Refactored attributes to avoid ZEND API changes.

nikic commented on Zend/zend_attributes.c:
```diff
@@ -84,9 +70,52 @@ ZEND_API zend_attribute *zend_get_parameter_attribute_str(HashTable *attributes,
+
+	zend_attribute *attr = pemalloc(ZEND_ATTRIBUTE_SIZE(argc), persistent);
+
+	if (persistent == (zend_bool) (GC_FLAGS(name) & IS_STR_PERSISTENT)) {
```
This won't do what you're expecting. You need an explicit != 0 check for the flag.

nikic commented on Zend/zend_attributes.h:
```diff
@@ -37,8 +35,59 @@ ZEND_API zend_attribute *zend_get_parameter_attribute_str(HashTable *attributes,
+	return zend_add_attribute(&info->attributes, ce->type != ZEND_USER_CLASS, 0, name, argc);
+}
+
+static zend_always_inline zend_attribute *zend_add_property_attribute_str(zend_class_entry *ce, const char *key, size_t len, zend_string *name, uint32_t argc)
```
I'd go for zend_string here.

nikic commented on Zend/zend_attributes.h:
```diff
@@ -37,8 +35,59 @@ ZEND_API zend_attribute *zend_get_parameter_attribute_str(HashTable *attributes,
+	return zend_add_attribute(&info->attributes, ce->type != ZEND_USER_CLASS, 0, name, argc);
+}
+
+static zend_always_inline zend_attribute *zend_add_property_attribute_str(zend_class_entry *ce, const char *key, size_t len, zend_string *name, uint32_t argc)
```
I'd accept a zend_string here instead, as the more flexible option.

kooldev replied:
I did not use zend_string because I thought the most common use case would be like this:
```c
zend_add_property_attribute_str(my_ce, ZEND_STRL("prop1"), attr_ce->name, 0);
zend_add_property_attribute_str(my_ce, ZEND_STRL("prop2"), attr_ce->name, 0);
```
Allocating a lot of strings seemed like more work for API users. But if we change the registration functions to return pointers to structures this would give easy access to the name as zend_string which would be even better.

nikic replied:
If we return pointers, there's probably no need for the _str APIs at all.

nikic commented on Zend/zend_compile.c:
```diff
@@ -6535,29 +6499,28 @@ void zend_compile_prop_decl(zend_ast *ast, zend_ast *type_ast, uint32_t flags, H
 		}
 
-		zend_declare_typed_property(ce, name, &value_zv, flags, doc_comment, attributes, type);
+		zend_declare_typed_property(ce, name, &value_zv, flags, doc_comment, type);
```
We can change the signature of this function to return the `zend_property_info *` instead.

nikic commented on Zend/zend_compile.c:
```diff
@@ -6609,7 +6566,12 @@ void zend_compile_class_const_decl(zend_ast *ast, zend_ast *attr_ast) /* {{{ */
 
 		zend_const_expr_to_zval(&value_zv, value_ast);
-		zend_declare_class_constant_ex(ce, name, &value_zv, ast->attr, doc_comment, attributes);
+		zend_declare_class_constant_ex(ce, name, &value_zv, ast->attr, doc_comment);
```
Similarly here, I'd return the pointer to the constant.

kooldev committed:
Addressed Review Comments.

kooldev commented:
@nikic I followed your suggestions and adjusted the code accordingly. I dropped the str-functions from the header, having the returned pointers is enough for attribute registration.

kooldev commented:
@beberlei Should be ready to merge now. ;-)
